op-node/rollup/derive/span_batch_txs.go

```diff
@@ -468,3 +468,40 @@ func (sbtx *spanBatchTxs) AddTxs(txs [][]byte, chainID *big.Int) error {
 	sbtx.totalBlockTxCount += totalBlockTxCount
 	return nil
 }
+
+func (sbtx *spanBatchTxs) revertLastBatch(txs []hexutil.Bytes) {
+	totalBlockTxCount := uint64(len(txs))
+	for idx := 0; idx < int(totalBlockTxCount); idx++ {
+		var tx types.Transaction
+		if err := tx.UnmarshalBinary(txs[len(txs)-1-idx] /* iterate in reverse order */); err != nil {
+			panic("impossible since already checked by AddTxs")
+		}
+
+		if tx.Type() == types.LegacyTxType {
+			// revert protectedBits and totalLegacyTxCount
+			sbtx.protectedBits.SetBit(sbtx.protectedBits, int(sbtx.totalLegacyTxCount)-1, 0)
+			sbtx.totalLegacyTxCount--
+		}
```

Comment:
I would really like some more concerted unit tests around this logic -- reversing the legacy tx bitmap (and all the other bitmaps) is a sensitive operation. I see we have a test for SpanBatchRevert, but since this is the more difficult workload, and could use its own tests.

I'll also think more about this logic, as I don't yet fully understand how it's working.

Reply:
It's basically symmetric reverse of `AddTxs`, the main caveat is that the reverted spanBatch can not be compared with its previous state directly, as two big.Int(s) of the same math value can have different representations. That's why in `TestSpanBatchRevert` I compare two spanBatch after serialization.